virttest/utils_test/libvirt.py

```diff
@@ -1918,23 +1918,43 @@ def create_hostdev_xml(pci_id, boot_order=None, xmlfile=True,
     return hostdev_xml.xml
 
 
-def create_controller_xml(contr_type='scsi', contr_model='virtio-scsi',
-                          contr_index='0', contr_alias=None):
+def create_controller_xml(vm_name, contr_dict, oper="get_xml"):
     """
-    Create controller xml
+    Create controller xml or add controller
 
-    :param contr_type: controller type name
-    :param contr_model: controller device model
-    :param contr_index: controller device index
-    :param contr_alias: controller device alias name
-    :return: controller xml file
-    """
+    :param vm_name: name of vm
+    :param contr_dict: The dict params includs controller configurations
+    :param oper: Operation need to do, current we have 2 choice
+                1. "get_xml": controller xml according to dict
+                2. "add_controller": add controller according to dict and updated in vm
+    :return: controller xml file if oper is equal to "get_xml";
+             controller object for 'add_controller'.
+    """
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    contr_type = contr_dict.get("controller_type", 'scsi')
+    contr_model = contr_dict.get("controller_model", "virtio-scsi")
+    contr_index = contr_dict.get("controller_index")
+    contr_alias = contr_dict.get("contr_alias")
     contr = controller.Controller(contr_type)
     contr.model = contr_model
-    contr.index = contr_index
+    if contr_index:
+        contr.index = contr_index
     if contr_alias:
         contr.alias = dict(name=contr_alias)
```

Comment:
I have new request here, need specify bus for controller, 
```
<controller index="0" model="virtio-transitional" type="scsi">
    <address **bus**="0x08" domain="0x0000" function="0x0" slot="0x01" type="pci" />
</controller>
```
Maybe something like:
```
address = contr.new_controller_address({'bus': controller_bus})
contr.address = address
```

Reply:
@vivianQizhu updated.